apis/python/src/tiledbsoma/fastercsx.cc

```diff
@@ -0,0 +1,621 @@
+            auto Bp_view = make_mutable_span_<CSX_MAJOR_INDEX>(Bp);
+            auto Bj_view = make_mutable_span_<CSX_MINOR_INDEX>(Bj);
+            auto Bd_view =
+                make_mutable_casted_span_<VALUE, remap_value_t<VALUE>>(Bd);
+
+            py::gil_scoped_release release;
+            return fastercsx::compress_coo(
+                ctx->thread_pool().get(),
+                shape,
+                nnz,
+                Ai_views,
+                Aj_views,
+                Ad_views,
+                Bp_view,
+                Bj_view,
+                Bd_view);
+        },
+        value_type,
+        csx_major_index_type,
+        csx_minor_index_type,
+        coo_index_type);
+}
+
+/**
+ * @brief python binding for sort_indices, which sorts minor dimension of a
+ * compressed matrix
+ */
+void sort_indices(
+    std::shared_ptr<tiledbsoma::SOMAContext> ctx,
+    py::array Bp,
+    py::array Bj,
+    py::array Bd) {
+    // Error checks first
+    //
+    if (Bp.ndim() != 1 || Bj.ndim() != 1 || Bd.ndim() != 1)
+        throw std::length_error("All arrays must be 1D");
+    if (!Bp.writeable() || !Bj.writeable() || !Bd.writeable())
+        throw std::invalid_argument("Output arrays must be writable.");
+
+    // Get dispatch types
+    CsxIndexType csx_major_index_type = lookup_dtype_(
+        csx_index_type_dispatch, Bp.dtype(), "CSx indptr array");
+    CsxIndexType csx_minor_index_type = lookup_dtype_(
+        csx_index_type_dispatch, Bj.dtype(), "CSx indices array");
+    ValueType value_type = lookup_dtype_(
+        value_type_dispatch, Bd.dtype(), "CSx data array");
+
+    // Dispatch by type
+    std::visit(
+        [&](auto value_type,
+            auto csx_major_index_type,
+            auto csx_minor_index_type) {
+            using CSX_MAJOR_INDEX =
+                typename decltype(csx_major_index_type)::type;
+            using CSX_MINOR_INDEX =
+                typename decltype(csx_minor_index_type)::type;
+            using VALUE = typename decltype(value_type)::type;
+
+            auto n_row = Bp.size() - 1;
+            int64_t nnz = py::cast<py::array_t<CSX_MAJOR_INDEX>>(Bp).at(n_row);
+            if (Bj.size() != nnz || Bd.size() != nnz)
+                throw std::length_error("Array length and nnz do not match.");
+
+            auto Bp_view = make_span_<CSX_MAJOR_INDEX>(Bp);
+            auto Bj_view = make_mutable_span_<CSX_MINOR_INDEX>(Bj);
+            auto Bd_view =
+                make_mutable_casted_span_<VALUE, remap_value_t<VALUE>>(Bd);
+
+            py::gil_scoped_release release;
+            return fastercsx::sort_indices(
+                ctx->thread_pool().get(),
+                n_row,
+                nnz,
+                Bp_view,
+                Bj_view,
+                Bd_view);
+        },
+        value_type,
+        csx_major_index_type,
+        csx_minor_index_type);
+};
+
+/**
+ * @brief Python binding for parallel slice+to_dense operation.
+ */
+void copy_to_dense(
```

Comment:
```suggestion
void copy_csx_to_dense(
```

Reply:
this is also an inconsistency if done to this one signature (why not also `sort_csx_indices`, etc?)

I will make the change, but don't love the introduction of inconsistent patterns of naming.

Reply:
As you wish -- this is code review and I'm offering suggestions from my point of view, for your consideration, which you're happy to take or leave.

My personal preference re

> why not also sort_csx_indices, etc?)

is to, yes, do so. (The function internally asserts that the input is CSR or CSC so, IMHO, putting that in function names seems like a sensible thing to do. And doing that in other similarly affected places IMO also makes sense.)

Reply:
Yeah, I don't like these types of redundant names when the context is entirely clear to the caller from the C++ namespace or Python package.  But I'll make the change.